### PR TITLE
Fixed to use copy api in multipart upload

### PR DESCRIPTION
**Pull request by ggtakec**
### Relevant Issue (if applicable)
#607 

### Details
Modified to use `Upload part` and `Copy part(api)` when multipart upload of object.

This will allow you to download and upload the smallest sized parts if you upload after writing a part of the object.
(In accordance with the minimum part size 5MB of multipart upload, useless download and upload will not be performed.)

As much as possible, s3fs will improve performance by uploading only the updated area.

The `nomixupload` option has been added to keep the previous behavior backward compatible.
When this option is specified, s3fs operates as before.
Also, when the nocopyapi, norenameapi and nomultipart options are specified, the same operation as before is performed.

Besides this, test cases for this modified are added. (and the two typo corrections.)

### CODE REVIEW

**Comment by gaul** [comment]
Sorry for my delayed response; I will look at this soon.

File: test/integration-test-main.sh [hunk 1]
```diff
@@ -352,6 +352,30 @@ function test_multipart_copy {
...
+
+    if [ `uname` = "Darwin" ]; then
+       cat /dev/null > $BIG_FILE
+    fi
+    dd if=/dev/urandom of="/tmp/${BIG_FILE}" bs=$BIG_FILE_LENGTH seek=0 count=1
+    dd if="/tmp/${BIG_FILE}" of="${BIG_FILE}" bs=$BIG_FILE_LENGTH seek=0 count=1
+
+    # modify directly(seek 7.5MB offset)
+    echo -n "0123456789ABCDEF" | dd of="${BIG_FILE}" bs=1 seek=7864320 conv=notrunc
+    echo -n "0123456789ABCDEF" | dd of="/tmp/${BIG_FILE}" bs=1 seek=7864320 conv=notrunc
```
Comment (gaul) [inline test/integration-test-main.sh]: This issues 16 1-byte writes -- do you intend this?

File: src/fdcache.cpp [hunk 2]
```diff
@@ -494,6 +502,117 @@ int PageList::GetUnloadedPages(fdpage_list_t& unloaded_list, off_t start, size_t
...
+//
+bool PageList::GetLoadPageListForMultipartUpload(fdpage_list_t& dlpages)
+{
+  // compress before this processing
+  if(!Compress()){
+    return false;
+  }
+
+  size_t    before_loaded_size = 0;
+  bool      before_indigent    = false;
```
Comment (gaul) [inline src/fdcache.cpp]: What does indigent mean?

**Comment by ggtakec** [comment]
@gaul I'm sorry about this, I will fix conflict and response your review comments.

File: src/fdcache.cpp (hunk 2 above)
Comment (ggtakec) [reply src/fdcache.cpp]: Sorry, it means "not enough of 5MB", I will change the flag name to "before_insufficient".

File: test/integration-test-main.sh (hunk 1 above)
Comment (ggtakec) [reply test/integration-test-main.sh]: Yes, I intend to issue 16 1-byte writes.
Would you like to write one 16 byte or something else?

**Comment by ggtakec** [comment]
@gaul 
> I did some further testing and found that the following multi-byte test fails:
> 
> ```shell
> dd if=/dev/zero of="${BIG_FILE}"      bs=4 count=1 seek=7864320 conv=notrunc
> dd if=/dev/zero of="/tmp/${BIG_FILE}" bs=4 count=1 seek=7864320 conv=notrunc
> ```
> 
> The logs suggest some bad input to S3:
> 
> ```
> s3fs: [ERR] curl.cpp:RequestPerform(2412): HTTP response code 400, returning EIO. Body Text: <?xml version='1.0' encoding='UTF-8'?><Error><Code>InvalidArgument</Code><Message>Bad Request</Message><RequestId>4442587FB7D0A2F9</RequestId></Error>
> ```
I was able to reproduce this bug.
The process of writing beyond the original file size seems wrong.

**Comment by ggtakec** [comment]
@gaul This PR is closed once for rebasing and generated as a new PR.
(Because rebase modification is various and you want to change the branch of the merge source)
I will close PR, so please wait.

File: test/integration-test-main.sh (hunk 1 above)
Comment (gaul) [reply test/integration-test-main.sh]: I think either is fine, but the test case below needs more than one byte to trigger a failure for some reason.

**Comment by ggtakec** [comment]
Now that the bug has been fixed, I will avoid the conflict , rebase and create a new PR.
I will close this PR now, but I will set the anchor from the new PR.

**Pull request closed without merging**